Stylable Toggle Switch

alisey opened the pull request:
The initial goal was to accept icons through props, but I noticed some issues, started changing stuff, and basically rewrote the component and added some adjustments to CSS and tests.

Changes:
* Fix type signatures
* Add props for icons and tabIndex
* Remove props for inline styles of internal elements
* Use the native checkbox to handle events
* Fix CSS for the focus outline (top: 50%; translateY(-50%) was giving the component incorrect bounding box)
* Move non-essential stying from the base style to theme
* Rename internal parts: outerLabel -> track, innerLabel -> knob, toggleIcon -> knobIcon
* Don't add autogenerated ID to the component (there's no point in it if it's private)
* Don't pass event to onChange handler (there's no need in it at the moment)
* Add focus and focusVisible states (the latter can be used to show the outline only when the component is focused via keyboard, like native checkboxes and material ui switches)

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,122 @@
+        <div className={tsStyle.track} />
+        <div className={tsStyle.knob}>
+          <div className={tsStyle.knobIcon}>
+            {checked ? this.props.checkedIcon : this.props.uncheckedIcon}
```
Did you check that it is ok also when there is no icon provided?

alisey replied:
Ok in what sense? If the icon is not provided it's just an empty div.

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,122 @@
+          </div>
+        </div>
         <input
+          id={this.props.id}
```
In this refactored component you don't need the id prop, so it is ok to remove it. There is no convention to have id prop for components

alisey replied:
I remember we needed this for Santa when I was working on pagination. Santa needs to be able to put ID on the root. Also useful for things like `<label for="myswitch">`

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,122 @@
+
+  // Doesn't get invoked if the input is disabled.
+  private handleMouseDown: React.MouseEventHandler<HTMLElement> = e => {
+    if (e.button === 0 && !this.state.focus) {
```
I am not sure I understand this condition.

alisey replied:
I'm going to remove the second part of it, it's no longer necessary. Would it be clear like this?
```
if (e.button === 0) {
    this.focusedByMouse = true;
}
```

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,122 @@
+  }
+
+  private handleFocus: React.FocusEventHandler<HTMLElement> = e => {
+    this.setState({focus: true, focusVisible: !this.focusedByMouse});
```
I think now that we have the new feature of state with params in stylable, we can have focus state which gets params "mouse" or "keyboard". It feels more semantically correct, instead of having focus & focusVisible   ----> focus(mouse) or focus(keyboard).
WDYT?

alisey replied:
Yeah, I like your suggestion, much more readable that way. On the other hand, `:focus-visible` is a part of CSS4 draft, if some day it becomes standard then I think it's preferable to follow the standard API. So I don't know really.

I'm going to go with your suggestion, because it's nicer, and we can reconsider in the future.

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.spec.tsx:
```diff
@@ -6,31 +6,29 @@ import {isTestkitExists} from 'wix-ui-test-utils/vanilla';
-      const driver = createDriver(<ToggleSwitch checked onChange={noop}/>);
-      expect(driver.isChecked()).toBeTruthy();
+      const driver = createDriver(<ToggleSwitch checked />);
+      expect(driver.isChecked()).toBe(true);
```
The difference between toBeTruthy() and toBe(true) is that toBeTruthy() returns true not only for "true" value, but also for any truthy value like 1, 'a', ... so semantically it is more correct, but It is not that important in this case

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.spec.tsx:
```diff
@@ -67,44 +65,28 @@ describe('ToggleSwitch', () => {
       expect(driver.getId()).toBe(testId);
     });
+
+    it('should apply user specified tabIndex', () => {
```
I would add a test "should have tabIndex 0  by default" as well

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.spec.tsx:
```diff
@@ -67,44 +65,28 @@ describe('ToggleSwitch', () => {
-      expect(driver.getId()).toBeDefined();
-    });
-
+  describe('attributes', () => {
```
Missing tests for the new configurable Icons feature. It is completely not tested. when checked. unchecked, w/o icons...

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.spec.tsx:
```diff
@@ -67,44 +65,28 @@ describe('ToggleSwitch', () => {
+      expect(isEnzymeTestkitExists(<ToggleSwitch />, enzymeToggleSwitchTestkitFactory, mount)).toBe(true);
     });
   });
 });
```
There are still some uncovered parts. For example, the focus new feature. You can test that the component is in the expected state when focusing with mouse or keyboard...

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.st.css:
```diff
@@ -1,34 +1,29 @@
-  cursor: pointer;
+.track {}
 
-  top: 50%;
```
This was crucial for santa, just pointing it out that we must be sure all desired features are still working for the product request

alisey replied:
Right, Gilad showed me why it's needed. But it shouldn't be part of the core, it should be in the stylable theme for santa toggle switch, which doesn't exist yet.

We also need to address this problem in ToggleSwitchStyle, and I plan to rewrite it, just don't want to expand the scope of this PR. You know what, I'm going to put this vertical alignment thing into ToggleSwitchStyle, which is going to break focus outline styling, but that's ok since it was already broken. And then I'm going to remove it and replace with something else later.
![image](https://user-images.githubusercontent.com/415252/36786825-763cac3c-1c90-11e8-9091-811a76e6ba33.png)

alisey commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,122 @@
+  }
+
+  private handleChange: React.ChangeEventHandler<HTMLInputElement> = e => {
+    if (this.props.onChange) {
```
I do want it optional, mainly because all event handlers for standard controls are optional. Will add it to default props.

alisey commented:
* Added inine styles back (incompatible with what Barak is doing, he removed them)
* onChange is now noop by default
* Added tests for the icon
* Added test for default tab index
* Added back `top: 50%` style

- Didn't change focus-visible to focus(keyboard/mouse), because then plain `:focus` stops working.
- Didn't add tests for focus behavior, because in node environment I would need to simulate the sequence of events I expect on mouse/keyboard press, which doesn't necessarily correspond to the sequence a browser would generate, and the tests would be passing even if the component is totally broken. We really need browser envoriment with non-simulated events for such tests.

shlomitc commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.tsx:
```diff
@@ -1,90 +1,128 @@
     return (
-      <label {...style('root', {checked, disabled}, this.props)} style={styles.root} tabIndex={0} ref={ref => this.toggle = ref}>
+      <div
+        {...tsStyle('root', {
```
why not just `style`? this is the convention around all other components

alisey replied:
There was `style`, which is inline style, then `styles`, which is a collection of styles for the parts of the component, and then also `style` which is the stylesheet. It was kinda confusing to me, and I wanted to make them more distinct.

shlomitc replied:
ok, so maybe we should give it a better name in general, but now it looks weird and it won't work for components with a single character like input. will it be `iStyle`? I even prefer calling it `s` instead.

alisey replied:
How about `css`? What we're importing is basically a stylesheet.

shlomitc replied:
cool

alisey replied:
Got a different idea. The imported style is now `style`, and the inline styles are `inlineStyles`.

lbelinsk commented on packages/wix-ui-core/src/components/StylableToggleSwitch/ToggleSwitch.st.css:
```diff
@@ -1,34 +1,25 @@
 .root {
-  -st-states: checked, disabled;
+  /* Info on focus-visible: https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo */
+  -st-states: checked, disabled, focus, focus-visible;
```
why in the end you decided not to use the stylable states with params feature?
Is it because of the CSS4?

alisey replied:
@lbelinsk That's one reason, another is that with `:focus(mouse/keyboard)` we lose the ability to use plain `:focus`. We had the same issue in checkbox, where we wanted to have `:checked(true/false/indeterminate)` and then went back to `:checked` and `:indeterminate`.
